docs/src/parameters.md
# Parameter Arguments
The following parameters can be used to as keyword arguments to modify a plot
## Plot parameters
These parameters modify the entire plot.

| Keyword     | Description | Default |
| ----------- | ----------- | ------- |
| `width`     | width of the plot in pixels | `500` |
| `height`    | height of the plot in pixels | `500` |
| `layout_algorithm` | algorithm for generating network layout (see [Layouts](@ref)) | `kamada_kawai` |
| `fixed` | use fixed coordinates from network model | `false` |
| `parallel_edge_offset` | offset distance between parallel edges | `0.05` |
| `node_components` | string or symbol array of node components to plot | `default_node_types`|
| `edge_components` | string or symbol array of edge components to plot | `default_edge_types`|
| `connected_components` | string or symbol array of connected components to plot | `default_connected_types`|

The current set of default component types includes:
```@example
using PowerPlots #hide
default_node_types
```
```@example
using PowerPlots #hide
default_edge_types
```
```@example
using PowerPlots #hide
default_connected_types
```

## Component Parameters
These parameters modify a specific component.

### Color
The `:color` argument can accept several inputs.  A single color can be specified using a color name as a symbol or a string.  [CSS color](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value) names are supported.  In addition, hex color values in a string are supported.

```julia
powerplot(case; branch=(:color=:yellow))
powerplot(case; branch=(:color="yellow"))
powerplot(case; branch=(:color="#FFA71A"))
```

A color range can be created by using several colors in an array. The range is used when component data is specified.

```julia
powerplot(case; branch=(:color=>[:red, "yellow", "#0000FF"]))
```

A color scheme from `ColorSchemes.jl` can be used, but the ColorScheme must be converted to an array of colors that can be interpreted by VegaLite.

```julia
using ColorSchemes
powerplot(case; branch=(:color=>colorscheme2array(ColorSchemes.colorschemes[:tableau_10])))
```

Default colors are determined by the order the components are plotted, shown in the legend.  The default color profiles are:

```@example
using PowerPlots, Colors # hide
parse.(Colorant, color_schemes[component_color_order[1]]) # hide
```
```@example
using PowerPlots, Colors # hide
parse.(Colorant, color_schemes[component_color_order[2]]) # hide
```
```@example
using PowerPlots, Colors # hide
parse.(Colorant, color_schemes[component_color_order[3]]) # hide
```
```@example
using PowerPlots, Colors # hide
parse.(Colorant, color_schemes[component_color_order[4]]) # hide
```
```@example
using PowerPlots, Colors # hide
parse.(Colorant, color_schemes[component_color_order[5]]) # hide
```
```@example
using PowerPlots, Colors # hide
parse.(Colorant, color_schemes[component_color_order[6]]) # hide
```
```@example
using PowerPlots, Colors # hide
parse.(Colorant, color_schemes[component_color_order[7]]) # hide
```
```@example
using PowerPlots, Colors # hide
parse.(Colorant, color_schemes[component_color_order[8]]) # hide
```
```@example
using PowerPlots, Colors # hide
parse.(Colorant, color_schemes[component_color_order[9]]) # hide
```
```@example
using PowerPlots, Colors # hide
parse.(Colorant, color_schemes[component_color_order[10]]) # hide
```
```@example
using PowerPlots, Colors # hide
parse.(Colorant, color_schemes[component_color_order[11]]) # hide
```
```@example
using PowerPlots, Colors # hide
parse.(Colorant, color_schemes[component_color_order[12]]) # hide
```
```@example
using PowerPlots, Colors # hide
parse.(Colorant, color_schemes[component_color_order[13]]) # hide
```
```@example
using PowerPlots, Colors # hide
parse.(Colorant, color_schemes[component_color_order[14]]) # hide
```
```@example
using PowerPlots, Colors # hide
parse.(Colorant, color_schemes[component_color_order[15]]) # hide
```
```@example
using PowerPlots, Colors # hide
parse.(Colorant, color_schemes[component_color_order[16]]) # hide
```

### Flow
Power flow can be shown on edges of the network.  For each edge component, the following parameters are available:

| Keyword | Description | Default |
| ------- | ----------- | ------- |
| `show_flow` | whether flow arrows representing 'pt' are displayed on edges | `false` |
| `show_flow_legend` | whether the legend for the flow arrows is shown | `false` |

```@example power_data
powerplot(case; branch=(:show_flow=>true, :show_flow_legend=>true))
```


### Size
The `:size` argument sets the size of a component.  The size does not vary with data in the base plot.

| Component     | Description | Default |
| ----------- | ----------- | ------- |
| `edge`     | default size of all edges except connector | `5` |
| `connector`    | default size of connector | `3` |
| `node`    |  set the size of all nodes | `500` |

```julia
powerplot(case; branch=(:size=>10), gen=(:size=>250))
```


The size of the flow arrows is a keyword that is a subset of the edge it models. However, the only first plotted component in the legend will use the `flow_arrow_size_range` keyword.  All subsequent components will use the same range as the first.
```julia
powerplot(case;
    branch=(:show_flow=>true, :show_flow_legend=>true,
     :flow_arrow_size_range=>[100,500]),
    switch=(:show_flow=>true, :show_flow_legend=>true,
     :flow_arrow_size_range=>[500,3000])
)
```


### Data
The `:data` argument selects the data from the component dictionary to use in the visualization.  The data argument can be a string or a symbol.  The data value modifies the color of a component based on the color range.

```julia
powerplot(case; gen(:data=:pmax))
powerplot(case; gen(:data=:pmin))
```


### Datatype
The datatypes in [VegaLite](https://vega.github.io/vega-lite/docs/type.html) can be `:nominal`, `:ordinal`, or `:quantintative`.  `:nominal` and `:ordinal` are both discrete values, and `:quantitative` is continuous.  In the context of the simple `powerplot`, there is no distinction between `:nominal` and `:ordinal`.  The default data type is `nominal`.

```julia
powerplot(case; gen(:data=:pg, :data_type=:quantitative)) # the pg is continous, so use a continous scale
powerplot(case; gen(:data=:index, :data_type=:nominal)) # the index is a discrete value, so use a discrete scale
```

### Other Parameters
The length of the dashes on the connector lines can be controlled with the `:dash` keyword.
The value is a two element array, with the first term representing the length of the gap and the second term representing the length of the dash.

| Keyword     | Description | Default |
| ----------- | ----------- | ------- |
| `:dash`     | set dash size for connectors | `[4,4]` |

```julia
powerplot(case, connector = (:dash=>[2,2]))
powerplot(case, connector = (:dash=>[2,8]))
```

CHANGELOG.md
PowerPlots.jl Change Log
========================
### Staged

### v0.5.1
- Add filter for hover tooltip
- Simplify geo-coordinate documentation example

### v0.5.0
- Change powerplot arguments for components (breaking)
- Add support for all PowerModels and PowerModelsDistribution components
- Add support for plotting new (non-standard) components.
- Fix bug in 'weights' used for layouts
- Re-export `save` from VegaLite
- Fixed documentation interactivity
- Bump Colors to v0.13
- Bump Geometry Basics to v0.5

### v0.4.6
- set stroke dash for connector
- Add parameter for selecting components to plot

### v0.4.5
- Add parameter to set flow arrow size
- Add parameter to set weight of different components in layouts
- Add `load` as supported component

### v0.4.4
- Add support for plotting parallel branches/dclines
- Add powerflow arrows to powerplots

### v0.4.3

### v0.4.2
- Add support for plotting 3-phase grid data
- Add support for multinetwork plots

### v0.3.0
- Change PowerModelsGraph specification
- Add documentation for PowerModelsGraph
- Update to DataFrames 1.0 (breaking)
- Add CompatHelper github action
- Add changelog.md


src/core/options.jl
# Default Color Schemes
color_schemes = Dict{Symbol,Any}(
    :blues => colorscheme2array(ColorSchemes.ColorScheme(range(Colors.colorant"#3182BD", Colors.colorant"#C6DBEF", length=5))),#["#$(Colors.hex(c))" for c in ColorSchemes.ColorScheme(range(Colors.colorant"#3182BD", Colors.colorant"#C6DBEF", length=5))],
    :greens => colorscheme2array(ColorSchemes.ColorScheme(range(Colors.colorant"#31A354", Colors.colorant"#C7E9C0", length=5))),
    :oranges => colorscheme2array(ColorSchemes.ColorScheme(range(Colors.colorant"#E6550D", Colors.colorant"#FDD0A2", length=5))),
    :reds => colorscheme2array(ColorSchemes.ColorScheme(range(Colors.colorant"#CB181D", Colors.colorant"#FCBBA1", length=5))),
    :purples => colorscheme2array(ColorSchemes.ColorScheme(range(Colors.colorant"#756BB1", Colors.colorant"#DADAEB", length=5))),
    :grays => colorscheme2array(ColorSchemes.ColorScheme(range(Colors.colorant"#555555", Colors.colorant"#FFFFFF", length=5))),
    :browns => colorscheme2array(ColorSchemes.ColorScheme(range(Colors.colorant"#8C6D31", Colors.colorant"#E7CB94", length=5))),
    :pinks => colorscheme2array(ColorSchemes.ColorScheme(range(Colors.colorant"#e7298a", Colors.colorant"#fbb3cc", length=5))),
    :yellows => colorscheme2array(ColorSchemes.ColorScheme(range(Colors.colorant"#ffc108", Colors.colorant"#ffef79", length=5))),
    :teal => colorscheme2array(ColorSchemes.ColorScheme(range(Colors.colorant"#2a8482", Colors.colorant"#46e1d1", length=5))),
    :lime => colorscheme2array(ColorSchemes.ColorScheme(range(Colors.colorant"#8df33c", Colors.colorant"#c8ff87", length=5))),
    :violet => colorscheme2array(ColorSchemes.ColorScheme(range(Colors.colorant"#8f3f8f", Colors.colorant"#d6a0d6", length=5))),
    :cyan => colorscheme2array(ColorSchemes.ColorScheme(range(Colors.colorant"#00ffff", Colors.colorant"#99ffff", length=5))),
    :magenta => colorscheme2array(ColorSchemes.ColorScheme(range(Colors.colorant"#ff00ff", Colors.colorant"#ff99ff", length=5))),
    :indigo => colorscheme2array(ColorSchemes.ColorScheme(range(Colors.colorant"#4b0082", Colors.colorant"#9b59b6", length=5))),
    :cherry => colorscheme2array(ColorSchemes.ColorScheme(range(Colors.colorant"#ff0000", Colors.colorant"#ff9999", length=5))),
)

component_color_order = Dict{Int,Symbol}(
    1 => :blues,
    2 => :greens,
    3 => :oranges,
    4 => :reds,
    5 => :purples,
    6 => :browns,
    7 => :pinks,
    8 => :yellows,
    9 => :lime,
    10 => :violet,
    11 => :cyan,
    12 => :magenta,
    13 => :indigo,
    14 => :cherry,
    15 => :teal,
    16 => :grays,
)


default_plot_attributes = Dict{Symbol,Any}(
    :width => 500,
    :height => 500,
    :parallel_edge_offset => 0.05,
    :node_components => default_node_types,
    :edge_components => default_edge_types,
    :connected_components => default_connected_types,
)

default_edge_attributes = Dict{Symbol,Any}(
    :color => nothing,
    :size => 5,
    :data => "ComponentType",
    :data_type => "nominal",
    :flow_color => :black,
    :flow_arrow_size_range => [500, 3000],
    :show_flow => false,
    :show_flow_legend => false,
    :hover => nothing
)

default_connector_attributes = Dict{Symbol,Any}(
    :color => [:gray],
    :size => 3,
    :dash => [4, 4],
    :data => "ComponentType",
    :data_type => "nominal",
    :hover => nothing
)

default_node_attributes = Dict{Symbol,Any}(
    :color => nothing,
    :size => 5e2,
    :data => "ComponentType",
    :data_type => "nominal",
    :hover => nothing
)

default_layout_attributes = Dict{Symbol,Any}(
    :fixed => false,
    :layout_algorithm => "spring",
    :connector_weight => 0.5,
    :edge_weight => 1.0,
    :node_weight => 1.0,
    :tol => 0.01,
    :iterations => 500,
    :K => 1.0,
    :C => 0.2,
    :initialtemp=>2.0,
    :nlist => [[1]],
    :abstols => 0.0,
    :reltols => 10e-6,
    :abstolx => 10e-6,
    :weights => Array{Float64}(undef, 0, 0),
    :uncon_dist=>1,
    :nodeweights => Float64[],
    :cols=>:auto,
    :dx=>1.0,
    :dy=>-1.0,
    :skip => Tuple{Int,Int}[],
)


const _color_attributes = [ # color (String or Symbol) type parameters
    :color,
]
const _numeric_attributes = [ # numeric parameters
    :size,
    :width,
    :height,
    :parallel_edge_offset,
]
const _label_attributes = [ # label (String or Symbol) type parameters
    :data,
    :data_type
]
const _boolean_attributes = [ # boolean type parameters
    :show_flow,
    :show_flow_legend,
]

